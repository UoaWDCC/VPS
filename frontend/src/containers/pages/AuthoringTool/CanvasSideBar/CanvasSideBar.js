import React from "react";

import styles from "../../../../styling/CanvasSideBar.module.scss";

export default function Canvas() {
  return (
    <>
      <div className={styles.sideBar} />
    </>
  );
}
