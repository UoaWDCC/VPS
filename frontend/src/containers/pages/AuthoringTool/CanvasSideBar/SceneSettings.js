import React, { useContext } from "react";
import TextField from "@material-ui/core/TextField";
import { withStyles } from "@material-ui/core/styles";
import { InputAdornment } from "@material-ui/core";
import SceneContext from "../../../../context/SceneContext";

import styles from "../../../../styling/CanvasSideBar.module.scss";

const CustomTextField = withStyles({
  root: {
    marginTop: "0.5em",
    marginBottom: "1.5em",

    "& label.Mui-focused": {
      color: "#0080a7 ",
    },
    "& .MuiInput-underline:after": {
      borderBottomColor: "0080a7",
    },
  },
})(TextField);

/**
 * This component displays the settings of a scene, such as the scene name
 * @component
 */
export default function SceneSettings() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);

  return (
    <>
      <div className={styles.sceneSettingsContainer}>
        <h1 className={styles.sideBarHeader}>Scene Settings</h1>
        <div className={styles.sideBarBody}>
          {/* input for scene name */}
          <CustomTextField
            label="Scene Name"
            value={currentScene?.name}
            fullWidth
            onChange={(event) => {
              setCurrentScene({
                ...currentScene,
                name: event.target.value,
              });
            }}
          />
          {/* input for scene timer duration */}
          <CustomTextField
            label="Scene Timer Duration"
            type="number"
            value={currentScene?.time || ""}
            fullWidth
            onChange={(event) => {
              // limiting scene timer duration
              const timeInput = event.target.value < 0 ? 0 : event.target.value;

              setCurrentScene({
                ...currentScene,
                time: timeInput,
              });
            }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">seconds</InputAdornment>
              ),
            }}
            InputLabelProps={{
              // label moves up whenever there is input
              shrink: currentScene?.time || currentScene?.time === 0,
            }}
          />
        </div>
      </div>
    </>
  );
}
