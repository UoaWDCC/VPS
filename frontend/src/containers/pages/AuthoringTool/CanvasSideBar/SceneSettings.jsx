import {
  InputLabel,
  InputAdornment,
  FormControl,
  FormControlLabel,
  Select,
  Checkbox,
  Typography,
} from "@material-ui/core";
import { useParams } from "react-router-dom";
import { usePut } from "hooks/crudHooks";
import AuthenticationContext from "context/AuthenticationContext";
import TextField from "@material-ui/core/TextField";
import { withStyles } from "@material-ui/core/styles";
import { useContext, useState, useEffect } from "react";
import ScenarioContext from "context/ScenarioContext";
import SceneContext from "../../../../context/SceneContext";

import styles from "../../../../styling/CanvasSideBar.module.scss";
import CustomInputLabelStyles from "./CustomPropertyInputStyles/CustomInputLabelStyles";
import CustomCheckBoxStyles from "./CustomPropertyInputStyles/CustomCheckBoxStyles";

const CustomInputLabel = CustomInputLabelStyles()(InputLabel);
const CustomCheckBox = CustomCheckBoxStyles()(Checkbox);
const CustomTextField = withStyles({
  root: {
    marginTop: "0.5em",
    marginBottom: "1em",

    "& label.Mui-focused": {
      color: "#0080a7 ",
    },
    "& .MuiInput-underline:after": {
      borderBottomColor: "0080a7",
    },
  },
})(TextField);

/**
 * This component displays the settings of a scene, such as the scene name
 * @component
 */
export default function SceneSettings() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);
  const { roleList } = useContext(ScenarioContext);
  const [selectedRoles, setSelectedRoles] = useState([]);

  const [checked, setChecked] = useState([]);
  const [allChecked, setAllChecked] = useState(false);

  useEffect(() => {
    if (currentScene.roles) {
      setSelectedRoles(currentScene.roles);
    }
    if (roleList.length > 0) {
      const initialCheckedState = roleList.map((role) =>
        (currentScene.roles || []).includes(role)
      );
      setChecked(initialCheckedState);

      // Initialize allChecked state
      const initialAllCheckedState = initialCheckedState.every(
        (check) => check
      );
      setAllChecked(initialAllCheckedState);
    }
  }, [roleList, currentScene.roles]);

  const { scenarioId } = useParams();
  const { scenes } = useContext(SceneContext);
  const { getUserIdToken } = useContext(AuthenticationContext);
  async function saveRoles(newRoles) {
    const updatedScenes = scenes.map(({ _id, name, roles: oldRoles }) => {
      if (_id === currentScene._id) {
        const roles = newRoles || oldRoles;
        return {
          _id,
          name,
          roles,
        };
      }
      return { _id, name, roles: oldRoles };
    });
    await usePut(
      `/api/scenario/${scenarioId}/scene/roles`,
      updatedScenes,
      getUserIdToken
    );
  }

  const handleCheckboxChange = (index) => {
    const newChecked = [...checked];
    newChecked[index] = !checked[index];
    setChecked(newChecked);

    const isAllChecked = newChecked.every((isChecked) => isChecked);
    setAllChecked(isAllChecked);

    const updatedSelectedRoles = newChecked.reduce((acc, isChecked, idx) => {
      if (isChecked) {
        acc.push(roleList[idx]);
      }
      return acc;
    }, []);
    setSelectedRoles(updatedSelectedRoles);
    saveRoles(updatedSelectedRoles);
  };

  const handleAllToggle = () => {
    const newAllChecked = !allChecked;
    setAllChecked(newAllChecked);

    const newChecked = newAllChecked
      ? new Array(roleList.length).fill(true)
      : new Array(roleList.length).fill(false);
    setChecked(newChecked);

    const newSelectedRoles = newAllChecked ? roleList : [];
    setSelectedRoles(newSelectedRoles);
    saveRoles(newSelectedRoles);
  };

  return (
    <>
      <div className={styles.sceneSettingsContainer}>
        <h1 className={styles.sideBarHeader}>Scene Settings</h1>
        <div className={styles.sideBarBody}>
          {/* input for scene name */}
          <CustomTextField
            label="Scene Name"
            value={currentScene?.name}
            fullWidth
            onChange={(event) => {
              setCurrentScene({
                ...currentScene,
                name: event.target.value,
              });
            }}
          />
          {/* input for scene timer duration */}
          <CustomTextField
            label="Scene Timer Duration"
            type="number"
            value={currentScene?.time ?? 0}
            fullWidth
            onChange={(event) => {
              // limiting scene timer duration
              const timeInput = event.target.value < 0 ? 0 : event.target.value;

              setCurrentScene({
                ...currentScene,
                time: timeInput,
              });
            }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">seconds</InputAdornment>
              ),
            }}
            InputLabelProps={{
              shrink: true,
            }}
          />
          <FormControl fullWidth className={styles.formControl}>
            <CustomInputLabel>Scene Role(s)</CustomInputLabel>
            <Select
              className={styles.selectInput}
              MenuProps={{
                getContentAnchorEl: null,
              }}
              multiple
              value={selectedRoles}
              onChange={(event) => setSelectedRoles(event.target.value)}
              renderValue={(selected) =>
                selected.length === roleList.length
                  ? "All"
                  : selected.join(", ")
              }
            >
              {roleList && roleList.length > 0 ? (
                <div style={{ display: "flex", flexDirection: "column" }}>
                  <FormControlLabel
                    control={
                      <CustomCheckBox
                        checked={allChecked}
                        onChange={handleAllToggle}
                      />
                    }
                    label="All"
                  />
                  {roleList.map((role, index) => (
                    <FormControlLabel
                      control={
                        <CustomCheckBox
                          checked={checked[index]}
                          onChange={() => handleCheckboxChange(index)}
                        />
                      }
                      label={role}
                    />
                  ))}
                </div>
              ) : (
                <Typography className={styles.menuItem}>
                  This scenario currently does not accommodate roles. Please
                  upload a CSV file first under Manage Groups.
                </Typography>
              )}
            </Select>
          </FormControl>
        </div>
      </div>
    </>
  );
}
