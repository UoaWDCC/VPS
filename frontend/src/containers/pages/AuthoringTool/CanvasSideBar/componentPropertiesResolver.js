import React from "react";
import ButtonPropertiesComponent from "./ComponentProperties/ButtonPropertiesComponent";
import TextPropertiesComponent from "./ComponentProperties/TextPropertiesComponent";

export default function componentPropertiesResolver(component, componentIndex) {
  switch (component?.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonPropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );
    case "TEXT":
      return (
        <TextPropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );
    default:
      break;
  }

  return <div />;
}
