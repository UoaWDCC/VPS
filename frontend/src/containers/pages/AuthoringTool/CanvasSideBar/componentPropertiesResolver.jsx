import ButtonPropertiesComponent from "./ComponentProperties/ButtonPropertiesComponent";
import FirebaseAudioPropertiesComponent from "./ComponentProperties/FirebaseAudioPropertiesComponent";
import ImagePropertiesComponent from "./ComponentProperties/ImagePropertiesComponent";
import SpeechTextPropertiesComponent from "./ComponentProperties/SpeechTextPropertiesComponent";
import TextPropertiesComponent from "./ComponentProperties/TextPropertiesComponent";
/**
 * This function returns the appropriate properties component for a scene component object
 *
 * @example
 * componentPropertiesResolver(currentScene.components[selectedComponentIndex], selectedComponentIndex)
 */
export default function componentPropertiesResolver(component, componentIndex) {
  switch (component?.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonPropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );

    case "SPEECHTEXT":
      return (
        <SpeechTextPropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );
    case "TEXT":
      return (
        <TextPropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );
    case "FIREBASEAUDIO":
      return (
        <FirebaseAudioPropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );
    case "FIREBASEIMAGE":
      return (
        <ImagePropertiesComponent
          component={component}
          componentIndex={componentIndex}
        />
      );
    default:
      break;
  }

  return <div />;
}
