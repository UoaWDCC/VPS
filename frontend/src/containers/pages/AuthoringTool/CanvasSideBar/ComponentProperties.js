import React, { useContext } from "react";
import AuthoringToolContext from "../../../../context/AuthoringToolContext";
import SceneContext from "../../../../context/SceneContext";
import componentPropertiesResolver from "./componentPropertiesResolver";

import styles from "../../../../styling/CanvasSideBar.module.scss";

export default function ComponentProperties() {
  const { select } = useContext(AuthoringToolContext);
  const { currentScene } = useContext(SceneContext);

  console.log(`id: ${select}`);
  console.log(currentScene.components);

  return (
    <>
      <div className={styles.componentPropertiesContainer}>
        <h1 className={styles.sideBarHeader}>Component Properties</h1>
        <div
          className={`${styles.sideBarBody} ${
            select === null ? styles.empty : ""
          }`}
        >
          {select === null ? (
            <p className={styles.noComponentSelectedText}>
              No component selected
            </p>
          ) : (
            componentPropertiesResolver(currentScene.components[select], select)
          )}
        </div>
      </div>
    </>
  );
}
