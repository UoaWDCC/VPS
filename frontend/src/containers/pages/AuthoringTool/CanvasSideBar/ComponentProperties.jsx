import { useContext } from "react";
import AuthoringToolContext from "../../../../context/AuthoringToolContext";
import SceneContext from "../../../../context/SceneContext";
import componentPropertiesResolver from "./componentPropertiesResolver";

import styles from "../../../../styling/CanvasSideBar.module.scss";

/**
 * This component displays the properties the selected scene component
 * @component
 */
export default function ComponentProperties() {
  const { select } = useContext(AuthoringToolContext);
  const { currentScene } = useContext(SceneContext);

  return (
    <>
      <div className={styles.componentPropertiesContainer}>
        <h1 className={styles.sideBarHeader}>Component Properties</h1>
        <div
          className={`${styles.sideBarBody} ${
            select === null ? styles.empty : ""
          }`}
        >
          {select === null ? (
            <p className={styles.noComponentSelectedText}>
              No component selected
            </p>
          ) : (
            componentPropertiesResolver(currentScene.components[select], select)
          )}
        </div>
      </div>
    </>
  );
}
