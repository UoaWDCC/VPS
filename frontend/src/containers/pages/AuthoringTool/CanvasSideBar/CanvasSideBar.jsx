import ComponentProperties from "./ComponentProperties";
import SceneSettings from "./SceneSettings";

import styles from "../../../../styling/CanvasSideBar.module.scss";

/**
 * This component displays the properties of scene components in a sidebar
 * @component
 */
export default function CanvasSideBar() {
  return (
    <>
      <div className={styles.sideBar}>
        <SceneSettings />
        <ComponentProperties />
      </div>
    </>
  );
}
