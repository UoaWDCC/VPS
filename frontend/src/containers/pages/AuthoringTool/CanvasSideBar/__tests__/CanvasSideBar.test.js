import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import CanvasSideBar from "../CanvasSideBar";
import ScenarioContextProvider from "../../../../../context/ScenarioContextProvider";
import SceneContextProvider from "../../../../../context/SceneContextProvider";

test("Scenario Selection page snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <ScenarioContextProvider>
        <SceneContextProvider>
          <CanvasSideBar />
        </SceneContextProvider>
      </ScenarioContextProvider>
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
