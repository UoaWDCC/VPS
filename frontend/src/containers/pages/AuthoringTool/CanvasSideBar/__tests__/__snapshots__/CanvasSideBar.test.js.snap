// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario Selection page snapshot test 1`] = `
<div
  className="sideBar"
>
  <div
    className="sceneSettingsContainer"
  >
    <h1
      className="sideBarHeader"
    >
      Scene Settings
    </h1>
    <div
      className="sideBarBody"
    >
      <div
        className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-1 MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Scene Name
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-1 MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Scene Timer Duration
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={0}
          />
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              seconds
            </p>
          </div>
        </div>
      </div>
      <div
        className="MuiFormControl-root formControl MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root WithStyles(ForwardRef(InputLabel))-root-2 MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Scene Role(s)
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline selectInput MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value=""
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="componentPropertiesContainer"
  >
    <h1
      className="sideBarHeader"
    >
      Properties
    </h1>
    <div
      className="sideBarBody empty"
    >
      <p
        className="noComponentSelectedText"
      >
        No component selected
      </p>
    </div>
  </div>
</div>
`;
